doc/debug.rst

```diff
@@ -0,0 +1,318 @@
+.. raw:: html
+  <blockquote><p>
+  The process of identifying and removing errors from computer hardware or software.
+  </p></blockquote>
+
+How to debug different subsystems of the compiler using built-in tooling.
+
+Special defines
+---------------
+
+Debugging functionality can usually be accessed only when compiler itself
+has been built with special defines - this is made to avoid runtime
+overhead, because some debugging tools are not exactly cheap to run.
+
+**Used when compiling temporary compiler**
+
+==================== =======
+Define               Enables
+-------------------- -------
+`nimVMDebugExecute`    Print out every instruction executed by the VM
+`nimVMDebugGenerate`   List VM code generated for every procedure called at compile-time
+`nimDebugUtils`        Enable semantic analysis execution tracer
+==================== =======
+
+**Used when executing temporary compiler**
+
+============================ =======
+Define                       Enables
+---------------------------- -------
+`nimCompilerDebug`           Reports for localized piece of the user code
+`nimCompilerDebugCalltrace`  Call trace reports
+`nimCompilerDebugTraceDir`   Write call traces to the directory
+============================ =======
+
+
+Semantic analysis
+-----------------
+
+When `nimDebugUtils` is enabled you can annotate section of the code to
+trace how semantic analysis is performed on this block of code. Each
+procedure that is annotated with one of the `addInNimDebugUtils()`
+overloads is going to be traced in the execution tree. For example -
+`sem.semOverloadedCall`, triggered when it is necessary to analyze call to
+overloaded procedure.
+
+.. code-block:: nim
+
+    proc semOverloadedCall(c: PContext, n, nOrig: PNode,
+                           filter: TSymKinds, flags: TExprFlags): PNode {.nosinks.} =
+      addInNimDebugUtils(c.config, "semOverloadedCall")
+      # Other implementaion parts ...
+
+If you compile your test file with `nim c -d:nimCompilerDebugCalltrace
+--filenames:canonical file.nim`:cmd: and annotate code block with the
+`nimCompilerDebug` wrapper.
+
+.. code-block:: nim
+
+    {.define(nimCompilerDebug).}
+    vmTarget(Obj())
+    {.undef(nimCompilerDebug).}
+
+You will get all the call entries traced
+
+.. important::
+
+     Both *compiler* and and your file must be compiled with defines - main
+     compiler requires `nimDebugUtils` (this guard exists to avoid heavy
+     performance hits), and your test compiler need to be run with
```

Comment:
no, it means that main compiler compiles tests compiler, there is no error. Previous suggestion removing `main` is also incorrect.

Reply:
I don't know how to word this properly, but considering we are compiling the compiler to compile the file there is bound to be some weird phrasing

Reply:
Then I need to follow through the instructions again, because I'm not getting the mechanics.

I would think we'd want to flip switches (defines) that would produce a test committee that'll output all the things. Then when we go to use that test compiler we provide more switches (defines) that would control the output.

Reply:
My understanding is:
1. Create a compiler that knows to trace (compile a test compiler with defines)
2. Use the test compiler created in the prior step to compile your program with tracing parameters to control the output